File: crates/defi-core/src/traits/yield_aggregator.rs
use alloy::primitives::{Address, U256};
use async_trait::async_trait;

use crate::error::Result;
use crate::types::*;

#[async_trait]
pub trait YieldAggregator: Send + Sync {
    fn name(&self) -> &str;
    async fn get_vaults(&self) -> Result<Vec<VaultInfo>>;
    async fn build_deposit(
        &self,
        vault: Address,
        amount: U256,
        recipient: Address,
    ) -> Result<DeFiTx>;
    async fn build_withdraw(
        &self,
        vault: Address,
        amount: U256,
        recipient: Address,
        owner: Address,
    ) -> Result<DeFiTx>;
}

File: ts/packages/defi-core/src/traits/yield_aggregator.ts
import type { Address } from "viem";
import type { VaultInfo, DeFiTx } from "../types.js";

export interface IYieldAggregator {
  name(): string;
  getVaults(): Promise<VaultInfo[]>;
  buildDeposit(vault: Address, amount: bigint, recipient: Address): Promise<DeFiTx>;
  buildWithdraw(vault: Address, amount: bigint, recipient: Address, owner: Address): Promise<DeFiTx>;
}
